firebase/functions/osm_auth.js

```diff
@@ -0,0 +1,238 @@
+                results,
+            );
+
+            // We have an OSM access token and the user identity now.
+            const accessToken = results.access_token;
+            if (accessToken === undefined) {
+                throw new Error(
+                    'Could not get an access token from OpenStreetMap',
+                );
+            }
+            // get the OSM user id and display_name
+            const { id, display_name } = await getOSMProfile(accessToken);
+            functions.logger.log('osmuser:', id, display_name);
+
+            // Create a Firebase account and get the Custom Auth Token.
+            const firebaseToken = await createFirebaseAccount(
+                id,
+                display_name,
+                accessToken,
+            );
+            functions.logger.log(
+                'fb token:',
+                firebaseToken,
+                'accessToken',
+                accessToken,
+            );
+            // build a deep link so we can send the token back to the app
+            // from the browser
+            const signinUrl = `${OAUTH_REDIRECT_URI}?token=${firebaseToken}`;
+            functions.logger.log('redirecting user to', signinUrl);
+            res.redirect(signinUrl);
+        });
+    } catch (error) {
+        // FIXME: this should show up in the user's browser as a bit of text
+        // We should figure out the various error codes available and feed them
+        // back into the app to allow the user to take action
+        return res.json({ error: error.toString() });
+    }
+});
+
+/**
+ * Creates a Firebase account with the given user profile and returns a custom
+ * auth token allowing the user to sign in to this account on the app.
+ * Also saves the accessToken to the datastore at v2/OSMAccessToken/$uid
+ *
+ * @returns {Promise<string>} The Firebase custom auth token in a promise.
+ */
+async function createFirebaseAccount(osmID, displayName, accessToken) {
+    // The UID we'll assign to the user.
+    // The format of these identifiers should NOT change once set
+    // The osmID is an integer which does not change for a given account,
+    // so the resulting firebase uid will look like `osm:123456`
+    // with a variable length.
+    const uid = `osm:${osmID}`;
+
+    // Save the access token to the Firebase Realtime Database.
+    const databaseTask = admin
+        .database()
+        .ref(`v2/OSMAccessToken/${uid}`)
+        .set(accessToken);
+
+    // Create or update the firebase user account.
+    // This does not login the user on the app, it just ensures that a firebase
+    // user account (linked to the OSM account) exists.
+    const userCreationTask = admin
+        .auth()
+        .updateUser(uid, {
+            displayName: displayName,
+        })
+        .catch(error => {
+            // If user does not exists we create it.
+            if (error.code === 'auth/user-not-found') {
+                return admin.auth().createUser({
```

Comment:
at what point in the workflow does the app write things to the user profile and when the firebase functions?

I think that currently the firebase functions do this when you login with OSM and the app does it when you log in with email + password. But I think that the app writes more data to firebase, e.g. the app saves the username to the user profile, the firebase functions don't do this. The app further displays a banner after successful login saying "welcome...", this is also not the case for the OSM login.

we could either check that both write the same information: https://github.com/mapswipe/mapswipe/blob/master/src/shared/views/Login.js#L203-L231

Reply:
another quick question here: https://dev-mapswipe.firebaseio.com/v2/OSMAccessToken/osm:undefined

this is probably an artifact from testing? how can the osm user id be `undefined`. I think that in such a case we will not create an acccount.

Reply:
Good catch on the `undefined` user. I've put a check in the function so it will error out in that case.
I have the same issue with the popup not being shown. I was convinced I had seen it when I first tested :thinking: 
As for updating the username, it looks like I missed the bit of code that does it.